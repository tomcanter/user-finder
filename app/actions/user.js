var Reflux = require('reflux');

module.exports = {
  searchUsername: Reflux.createAction()
};
