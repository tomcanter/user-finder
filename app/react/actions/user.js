import Reflux from 'reflux';
import request from 'reqwest';
import { getItem } from 'mixins/cache';
import gitHubAPI from 'api/github';
import partial from 'lodash-node/modern/function/partial'

const API_ROOT_URL = 'https://api.github.com';

var User = Reflux.createActions({
  search: { asyncResult: true },
  profile: { asyncResult: true },
  repos: { asyncResult: true },
  starred: { asyncResult: true }
});

User.search.listen(function(url) {
  var cachedData = getItem(url);

  if (cachedData) {
    return this.completed(cachedData, true);
  }

  gitHubAPI(url)
    .then(this.completed)
    .fail(partial(this.failed, url));
});

function fetchData(cacheKey, url, username) {
  var cachedData = getItem(`${cacheKey}:${username}`);

  if (cachedData) {
    return this.completed(cachedData, true);
  }

  request({
    url,
    type: 'json'
  })
    .then(this.completed)
    .fail(this.failed);
}

User.profile.listen(function(username) {
  fetchData.call(this, 'profile', `https://api.github.com/users/${username}`, username);
});
User.repos.listen(partial(fetchData, 'repos'));
User.starred.listen(partial(fetchData, 'starred'));

export default User
