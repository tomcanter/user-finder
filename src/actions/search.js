// @flow

import pick from 'lodash/fp/pick';
import flow from 'lodash/fp/flow';
import get from 'lodash/fp/get';
import mapValues from 'lodash/fp/mapValues';
import GitHub from 'github-api';
import {normalize} from 'normalizr';
import userListSchema from '../schema';

const gh = new GitHub();

const getUserState = get('entities.users');
const pickUserData = flow(
  getUserState,
  mapValues(
    pick([
      'login',
      'id',
      'avatar_url',
    ])
  )
);

export function searchUser({query}: {query: string}) {
  return function (dispatch: Function) {
    dispatch(searchRequest({query}));

    return gh
      .search({q: query})
      .forUsers()
      .then((response) => {
        const normalizedData = normalize(response.data, userListSchema);
        // Only take the properties of each user that are needed
        const entities = {
          users: pickUserData(normalizedData),
        };
        normalizedData.entities = entities;
        return dispatch(searchSuccess(query, normalizedData));
      })
      .catch(err => dispatch(searchFailure(err)));
  };
}

export const SEARCH_REQUEST: string = 'SEARCH_REQUEST';
export function searchRequest({query}: {query: string}) {
  return {
    type: SEARCH_REQUEST,
    query,
  };
}

export const SEARCH_SUCCESS: string = 'SEARCH_SUCCESS';
export function searchSuccess(query: string, data: Object) {
  const {result, entities} = data;
  return {
    entities,
    query,
    totalResults: result.length,
    type: SEARCH_SUCCESS,
    result,
  };
}

export const SEARCH_FAILURE: string = 'SEARCH_FAILURE';
export function searchFailure(error: Object) {
  const {response, message} = error;
  return {
    type: SEARCH_FAILURE,
    response,
    message,
  };
}
