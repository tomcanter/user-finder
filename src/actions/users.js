// @flow

import map from 'lodash/fp/map';
import pick from 'lodash/fp/pick';

export const USER_SEARCH_REQUEST: string = 'USER_SEARCH_REQUEST';
export function userSearchRequest({searchTerm}: {searchTerm: string}) {
  return {
    type: USER_SEARCH_REQUEST,
    searchTerm,
  };
}

const pickUserData = map(
  pick([
    'login',
    'id',
    'avatar_url',
  ])
);
export const USER_SEARCH_SUCCESS: string = 'USER_SEARCH_SUCCESS';
export function userSearchSuccess(response: Object) {
  const {data} = response;
  return {
    type: USER_SEARCH_SUCCESS,
    users: pickUserData(data),
    totalResults: data.length,
  };
}

export const USER_SEARCH_FAILURE: string = 'USER_SEARCH_FAILURE';
export function userSearchFailure(error: Object) {
  const {response, message} = error;
  return {
    type: USER_SEARCH_FAILURE,
    response,
    message,
  };
}
