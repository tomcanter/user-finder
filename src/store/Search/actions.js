// @flow

import get from 'lodash/fp/get';
import {normalize} from 'normalizr';
import assignAll from 'lodash/fp/assignAll';
import userSchema from '../schema';

function normalizeResponse(response: Object) {
  const normalized = normalize(response.items, userSchema);
  const {total_count, pagination} = response;
  return assignAll([
    normalized,
    {
      totalResults: total_count,
      pagination,
    },
  ]);
}

export function searchUser({query}: {query: string}) {
  return function (dispatch: Function, getState: Function, api: Object) {
    const cachedResult = get('search.cache', getState())[query];

    if (cachedResult) {
      return Promise
        .resolve()
        .then(() => dispatch(searchSuccess(cachedResult, cachedResult.query)));
    }

    dispatch(searchRequest({query}));

    return api
      .searchUsers({q: query})
      .then(normalizeResponse)
      .then(response => dispatch(searchSuccess(response, query)))
      .catch(err => dispatch(searchFailure(err)));
  };
}

export const SEARCH_REQUEST: string = 'SEARCH_REQUEST';
export function searchRequest({query}: {query: string}) {
  return {
    type: SEARCH_REQUEST,
    query,
  };
}

export const SEARCH_SUCCESS: string = 'SEARCH_SUCCESS';
export function searchSuccess(response: Object, query: string) {
  return assignAll([
    response,
    {
      type: SEARCH_SUCCESS,
      query,
    },
  ]);
}

export const SEARCH_FAILURE: string = 'SEARCH_FAILURE';
export function searchFailure(error: Object) {
  const {response, message} = error;
  return {
    type: SEARCH_FAILURE,
    response,
    message,
  };
}
