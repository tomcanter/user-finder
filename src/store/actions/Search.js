// @flow

import assignAll from 'lodash/fp/assignAll';

export function searchForUser(search: string) {
  return {
    type: 'SEARCH_REQUEST',
    search,
  };
}

export function searchSuccess(response: Object, query: string) {
  return assignAll([
    response,
    {
      type: 'SEARCH_SUCCESS',
      query,
    },
  ]);
}

export function searchFailure(error: Object) {
  const {response, message} = error;
  return {
    type: 'SEARCH_FAILURE',
    response,
    message,
  };
}
