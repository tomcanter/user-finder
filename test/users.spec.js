import * as actions from '../src/actions/users';

describe('when requesting users via a search term', () => {
  it('should dispatch a USER_SEARCH_REQUEST action', () => {
    expect(
      actions.userSearchRequest({searchTerm: 'tom'})
    ).toMatchSnapshot();
  });
});

describe('when receiving a set of users', () => {
  it('should dispatch a USER_SEARCH_SUCCESS action with the necessary data', () => {
    const users = [
      {login: 'foo', id: 123, other: 'test'},
      {login: 'baz', id: 456, other: 'test'},
    ];
    expect(
      actions.userSearchSuccess({data: users})
    ).toMatchSnapshot();
  });
});
