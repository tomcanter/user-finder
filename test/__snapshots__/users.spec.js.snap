exports[`when receiving a set of users should dispatch a USER_SEARCH_SUCCESS with only necessary data 1`] = `
Object {
  "totalResults": 2,
  "type": "USER_SEARCH_SUCCESS",
  "users": Array [
    Object {
      "id": 123,
      "login": "foo",
    },
    Object {
      "id": 456,
      "login": "baz",
    },
  ],
}
`;

exports[`when requesting users via a search term should dispatch a USER_SEARCH_REQUEST action 1`] = `
Object {
  "searchTerm": "tom",
  "type": "USER_SEARCH_REQUEST",
}
`;
